File: my-leetcode-solutions/maximum-depth-of-binary-tree.ts
// My solution to 104 - Maximum Depth of Binary Tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function maxDepth(root: TreeNode | null): number {
    // Base case
    if (root === null) {
        return 0;
    }

    const left = maxDepth(root.left) + 1
    const right = maxDepth(root.right) + 1
    return left > right ? left : right;
  

};

// Time Complexity: O(n)

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null

    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

const rootTestCases = [
    new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7))),
    new TreeNode(1, null, new TreeNode(2)),
    new TreeNode(1, new TreeNode(2)),
];

for (let i = 0; i < rootTestCases.length; i++) {
    const output = maxDepth(rootTestCases[i]);
    //  const output = preOrderTraversal(rootTestCases[i]);
    console.log(`Output: ${output} for root as ${JSON.stringify(rootTestCases[i])}`);
}

export {};


// traverse the tree using depth first search. when you get to the root, calculate the depth. if it's greater than max depth, change max depth to that height


// want to do pre order traversal

File: my-leetcode-solutions-py/maximum-depth-of-binary-tree.py
# My solution to 104 - Maximum Depth of Binary Tree

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root: TreeNode | None) -> int:
        # base case
        if root == None:
            return 0

        # recursive case
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))



# Time Complexity:

root_test_cases = [
    [3, 9, 20, None, None, 15, 7],
    [1, None, 2],
]

solution = Solution()

for i in range(len(root_test_cases)):
    nodes = [
        TreeNode(value) if value is not None else None
        for value in root_test_cases[i]
    ]
    children = iter(nodes[1:])

    for node in nodes:
        if node is not None:
            node.left = next(children, None)
            node.right = next(children, None)

    root = nodes[0]
    output = solution.maxDepth(root)
    print(f"Output: {output!r} for root as {root_test_cases[i]!r}")
